### PR TITLE
Show only the useful lines of stack trace - fixes #22

### LINKED ISSUE #22: Error context
Tape has this nice output on errors:

```
at: Test.<anonymous> (/Users/sindresorhus/dev/hypot/test.js:9:4)
```

Currently we output the full stack trace or none.

-

This might be something `claim` should handle.

https://github.com/substack/tape/blob/ccfcf4f76581abc3842163dfe065198764247c85/lib/test.js#L207

### PR DESCRIPTION
**Pull request by mdibaiee**
Fix "Assertion count does not match planned" error's stack trace not showing

### CODE REVIEW

**Comment by mdibaiee** [comment]
About promises, I couldn't find a way of finding the stack trace for them, and as it's self-explainatory that the the _returned promise_ was rejected, I think we can just ignore printing and stack trace for that.

**Comment by Qix-** [comment]
Without looking too deeply at the code, is there an option to show full stacks? This is one thing about Mocha I actually hate; a lot of the time I'm writing two modules that complement each other concurrently, and the stack traces never show the other's stack entries unless I use `--full-trace`.

**Comment by sindresorhus** [comment]
@Qix- What would you like to keep?

Currently it looks like:

```
  AssertionError: false === true
    at Test.(anonymous function) [as assert] (/Users/sindresorhus/dev/ava/lib/test.js:49:14)
    at Test.fn (/Users/sindresorhus/dev/acosh/test.js:8:4)
    at Test.run (/Users/sindresorhus/dev/ava/lib/test.js:78:18)
    at Runner.<anonymous> (/Users/sindresorhus/dev/ava/lib/runner.js:39:8)
    at Immediate.immediate._onImmediate (timers.js:418:16)
    at processImmediate [as _immediateCallback] (timers.js:367:17)
```

With this PR, it looks like:

```
false === true
    at Test.fn (/Users/sindresorhus/dev/acosh/test.js:8:4)
```

All it hides is useless stack info coming from AVA and Node.js core libs.

File: lib/test.js [hunk 1]
```diff
@@ -99,14 +101,15 @@ Test.prototype.end = function () {
 
 Test.prototype.exit = function () {
 	if (this.planCount !== null && this.planCount !== this.assertCount) {
+		console.log(this.planStack);
```
Comment (sindresorhus) [inline lib/test.js]: Leftover

**Comment by Qix-** [comment]
I'd have to see a full stack. If it's just removing lines caused by the tests then that should be just fine.

**Comment by sindresorhus** [comment]
Can you provide some code that result in a "full stack"? I definitely agree it should not remove that.

**Comment by Qix-** [comment]
Well more specifically with Mocha in general, it [removes any stack entries that have `node_modules` in them](https://github.com/mochajs/mocha/blob/2f458ab38cd191c4071742caa3d101a96a832c2b/lib/utils.js#L684-L740). I suppose it makes sense, but at the same time it's somewhat futile and time-wasting because, in the case of errors with suppressed `node_modules` stack lines, the error message doesn't normally correspond with the actual error site at the "top" of the trace, causing confusion.

**Comment by mdibaiee** [comment]
I see that this approach doesn't work in some cases, consider this example:

```javascript
test('something', function(t) {
  t.is('hey', 'hey');
  doSomeTest(t, 2);
  doSomeTest(t, {});
  t.end();
});

function doSomeTest(t, x) {
  t.same({x: x}, {x: {}});
}
```

This is the output for this example:
```
1. something
  { x: 2 } === { x: {} }
    at doSomeTest (/Users/mahdi/Documents/Workshop/github/ava-test/test.js:13:5)
```
But we should also show the line for:
```
at Test.fn (/Users/mahdi/Documents/Workshop/github/ava-test/test.js:9:5)
```

I'm going to see if I can find a better approach.

**Comment by Qix-** [comment]
Not quite, but that is true. For instance,

```javascript
// mocha

// Note, `my-module' is not part of the source tree, but it's still mine
// in this example. Maybe a sister/plugin module, etc.
var myModule = require('my-module');

it('should xyz', function () {
  myModule(); // throws
});
```

stack:
```
Error: some sort of error
    at function (test.js:8:4)
    at <mocha stuff>
```

It doesn't show where inside the module `my-module` the error was thrown. I have to pass `--full-trace` in order to get that information.

**Comment by mdibaiee** [comment]
Okay, I think I found the ultimate solution ✊

We can filter out the stack lines which start with the directory of the test file, see:

```javascript
let dir = path.dirname(module.parent.filename);

var split = (result.error.stack || '').split('\n');
let related = split.filter(line => line.indexOf(dir) > -1);

var beautiful = result.error.message + '\n' + related.join('\n');
```

**Comment by Qix-** [comment]
Then assertions won't show up correctly. :tangerine:

**Comment by mdibaiee** [comment]
Why? I tested it with the example I said above, it's working correctly, have you tested this?

Output:
```
0:50 » ~/Documents/Workshop/github/ava-test $ node test.js                                                                                                                                            ‹1›
  ✖ something { x: 2 } === { x: {} }

  1 test failed

  1. something
  { x: 2 } === { x: {} }
    at doSomeTest (/Users/mahdi/Documents/Workshop/github/ava-test/test.js:13:5)
    at Test.fn (/Users/mahdi/Documents/Workshop/github/ava-test/test.js:7:3)
```

**Comment by mdibaiee** [comment]
@sindresorhus Do you think I should squash my commits into one?

**Comment by sindresorhus** [comment]
@mdibaiee No need. I squash one merge. Easier to see individual changes this way.

**Comment by mdibaiee** [comment]
What's the plan now? Anything else has to be done in this PR?

**Comment by sindresorhus** [comment]
Looks good to me now :)

@Qix- ?

File: index.js [hunk 2]
```diff
@@ -36,6 +37,18 @@ function stack(results) {
 
 		i++;
 
+		// Don't print the full stack but the only useful line showing
+		// the actual test file stack
+		var dir = path.dirname(module.parent.filename);
```
Comment (Qix-) [inline index.js]: Sorry for my thickness, but what is this *actually* doing? Isn't it just filtering out anything that's not in the project directory? What is this improving/fixing?

Comment (mdibaiee) [inline index.js]: This is fixing [this](https://github.com/sindresorhus/ava/issues/22).

**Comment by Qix-** [comment]
I understand the desire for clean stack traces. I just heavily question the usefulness.

Tape's stack errors are even more useless than Mocha's if all they're showing is the relevant call site of the error. Many times, the parts of the stack that make the error's cause **immediately clear** are the parts above (deeper) the deepest user script. For example:

```
TypeError: Cannot convert object to primitive value
    at String (unknown source)
    at util.js:38:25
    at String.replace (native)
    at Object.<anonymous> (util.js:35:23)
    at Object.<anonymous> (console.js:25:36)
    at EventEmitter.<anonymous> (/project/src/routines/debug/boot.js:16:21)
```

Without looking at `boot.js` in the above project, I can safely assume the affected line is probably not clear what exactly is going on. Take a look.

```javascript
'use strict';

var path = require('path');

var message = 'TypeError: Cannot convert object to primitive value';
var stack = [
  '    at String (unknown source)',
  '    at util.js:38:25',
  '    at String.replace (native)',
  '    at Object.<anonymous> (util.js:35:23)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at EventEmitter.<anonymous> (' + __filename + ':16:21)',
].join('\n');

var dir = path.dirname(module.filename);
console.log('DIR', dir);

var split = (stack || '').split('\n');
var related = split.filter(function (line) {
	return line.indexOf(dir) > -1;
});

var beautiful = message + '\n' + related.join('\n');

console.log(beautiful);
```

which outputs

```
$ node test.js
DIR /private/tmp/avatest
TypeError: Cannot convert object to primitive value
    at EventEmitter.<anonymous> (/private/tmp/avatest/test.js:16:21)
```

I'd have no idea what was going on, making the stack trace **absolutely useless** - hence why I asked if there was a way to show full traces.

---

If the intent here is to only show relevant stack lines, so that we exclude:

- Lines that are part of Node.js proper
- Lines that belong to the AVA module

then why not filter on lines that have absolute paths, starting at the first found? The logic here is that, beginning at the deepest source file (going upward through the stack), we filter out anything that isn't a user script. We also show anything deeper than the deepest user script.

This is an approach I've used elsewhere for clean error printing and I've never run into problems with not seeing enough information.

```regex
\((?:\/(?:(?!node_modules[\/\\]ava[\/\\])[^:\/])+)+:\d+:\d+\)
```

Example:

```javascript
'use strict';

var path = require('path');

var message = 'TypeError: Cannot convert object to primitive value';
var stack = [
  '    at String (unknown source)',
  '    at util.js:38:25',
  '    at String.replace (native)',
  '    at Object.<anonymous> (util.js:35:23)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at EventEmitter.<anonymous> (' + __filename + ':16:21)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at EventEmitter.<anonymous> (' + __filename + ':16:21)',
  '    at EventEmitter.<anonymous> (' + __filename + ':16:21)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at Object.<anonymous> (console.js:25:36)'
].join('\n');

var reg = /\((?:\/(?:(?!node_modules[\/\\]ava[\/\\])[^:\/])+)+:\d+:\d+\)/;
function beautifulStack(stack) {
  var found = false;
  return stack.split('\n').filter(function (line) {
    var relevant = reg.test(line);
    found = found || relevant;
    return !found || relevant;
  }).join('\n');
}

console.log(message + '\n' + beautifulStack(stack));
```

which yields

```
$ node test.js
TypeError: Cannot convert object to primitive value
    at String (unknown source)
    at util.js:38:25
    at String.replace (native)
    at Object.<anonymous> (util.js:35:23)
    at Object.<anonymous> (console.js:25:36)
    at EventEmitter.<anonymous> (/private/tmp/avatest/test.js:16:21)
    at EventEmitter.<anonymous> (/private/tmp/avatest/test.js:16:21)
    at EventEmitter.<anonymous> (/private/tmp/avatest/test.js:16:21)
```

For @sindresorhus's stack example, this yields

```
$ node test.js
AssertionError: false === true
    at Test.(anonymous function) [as assert] (/Users/sindresorhus/dev/ava/lib/test.js:49:14)
    at Test.fn (/Users/sindresorhus/dev/acosh/test.js:8:4)
    at Test.run (/Users/sindresorhus/dev/ava/lib/test.js:78:18)
    at Runner.<anonymous> (/Users/sindresorhus/dev/ava/lib/runner.js:39:8)
```

---

Another example, [this stacktrace](http://stackoverflow.com/a/22726190/510036):

```
Error: `foo` has been removed in favorof `bar`
    at function() (/home/ray/dev/test/error.js:53:3)
    at b() (domain.js:183:18)
    at Domain.run() (domain.js:123:23)
    at function() (/home/ray/dev/test/error.js:52:3)
    at Module._compile() (module.js:456:26)
    at Module._extensions..js() (module.js:474:10)
    at Module.load() (module.js:356:32)
    at Module._load() (module.js:312:12)
    at Module.runMain() (module.js:497:10)
```

gets reduced down to

```
$ node test.js
Error: `foo` has been removed in favorof `bar`
    at function() (/home/ray/dev/test/error.js:53:3)
    at function() (/home/ray/dev/test/error.js:52:3)
```

That seems much more appropriate to me.

**Comment by mdibaiee** [comment]
@Qix- Can you give me the actual test that produces this stack?

```
at String (unknown source)',
  '    at util.js:38:25',
  '    at String.replace (native)',
  '    at Object.<anonymous> (util.js:35:23)',
  '    at Object.<anonymous> (console.js:25:36)',
  '    at EventEmitter.<anonymous> (' + __filename + ':16:21)',
```

**Comment by Qix-** [comment]
@mdibaiee No, it's an example stack trace I found on SO. My point is is that you don't need to look at the code to know that the first 5 frames are important for debugging the error.

**Comment by mdibaiee** [comment]
@Qix- My point is knowing what *your fault* was in the error is good enough for debugging most codes, and *starting* debugging of others.

**Comment by Qix-** [comment]
@mdibaiee aren't tests a vital part of debugging code?

**Comment by mdibaiee** [comment]
@Qix- You know the answer, yes. And how does that relate?

**Comment by Qix-** [comment]
Because a first glance view of all of the *relevant* parts of an error are important.

**Comment by mdibaiee** [comment]
I don't think so. I don't understand your worries, almost all commands have a `--verbose` option, do you know why? Because I don't need to know all the details everytime, I don't want `ls` to be like `ls -l` everytime, it takes spaces, it takes my time and my attention.

If you really think all relevant paths are important, `alias ava='ava --full-trace'`. Problem solved. I'm not continuing this discussion.

**Comment by Qix-** [comment]
Which is what I requested to begin with, however

```
$ grep -r "full-trace" *
``

You don't have the option in there.

**Comment by Qix-** [comment]
I'd also recommend setting infinity length stack traces before filtering.

```javascript
Error.stackTraceLimit = Infinity;
```

**Comment by sindresorhus** [comment]
Minimal vs verbose is a hard one, but from experience with debugging a lot of failed assertions I have to agree with @Qix-. I often have to digg deeper than just the test file. I'm willing to sacrifice visual beauty in favor of practicality here. We should preserve any stack info referencing user code as @Qix- has outlined above. I'm allergic to options and I'd rather go with a good default. We can always change it in the future if it turns out to be the wrong choice.

@mdibaiee Would you be willing to update your PR even though you seem to strongly disagree?

> Error.stackTraceLimit = Infinity;

:+1:

**Comment by Qix-** [comment]
One last thing (I know I'm persistent) but since the introduction of [`error-ex`](/Qix-/node-error-ex) we might run into exceptions with non-standard lines that start with something other than `at`; we shouldn't filter those out either.

**Comment by sindresorhus** [comment]
> but since the introduction of error-ex we might run into exceptions with non-standard lines that start with something other than at; we shouldn't filter those out either.

:+1: I've seen other error modules add stuff too. Better to be on the safe side.

**Comment by mdibaiee** [comment]
@sindresorhus So what's the plan now? Don't filter anything out? Use @Qix- 's RegExp?

**Comment by sindresorhus** [comment]
> Use @Qix- 's RegExp?

Yes

**Comment by mdibaiee** [comment]
@sindresorhus Alright, sorry for the delay, I was busy working on another projet, are we ready to go?

File: index.js [hunk 3]
```diff
@@ -36,6 +36,19 @@ function stack(results) {
 
 		i++;
 
+		// Don't print the full stack but the only useful line showing
+		// the actual test file stack
+		var reg = /\((?:[\\\/](?:(?!node_modules[\\\/]ava[\\\/])[^:\\\/])+)+:\d+:\d+\)/;
```
Comment (Qix-) [inline index.js]: Mind updating the regex?

```
(?:^(?! {4}at).{6})|(?:\((?:[\\\/](?:(?!node_modules[\\\/]ava[\\\/])[^:\\\/])+)+:\d+:\d+\))
```

This won't kill the error message and won't filter out lines that don't start with `at`.

File: index.js [hunk 4]
```diff
@@ -36,6 +36,19 @@ function stack(results) {
...
+		function beautifulStack(stack) {
+			var found = false;
+			return stack.split('\n').filter(function (line) {
+				var relevant = reg.test(line);
+				found = found || relevant;
+				return !found || relevant;
+			}).join('\n');
+		}
+		result.error.stack = beautifulStack(result.error.stack);
```
Comment (Qix-) [inline index.js]: Something makes me feel iffy about re-assigning it back to the error itself. Thoughts @sindresorhus?

Comment (sindresorhus) [inline index.js]: Agreed. Just use it inline below:

```js
log.writelpad(chalk.red(beautifulStack(result.error.stack)));
```

**Comment by mdibaiee** [comment]
@sindresorhus done

**Comment by sindresorhus** [comment]
Landed! Thanks for the PR @mdibaiee. Keep'em coming ;)

And thanks for the review @Qix-.

**Comment by Qix-** [comment]
No problem! Thanks @mdibaiee :dancer: :+1:

**Comment by mdibaiee** [comment]
✊✌

**Pull request closed without merging**